vendor/k8s.io/kubernetes/cmd/kubelet/app/patch_volumequota.go

```diff
@@ -0,0 +1,113 @@
+package app
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/golang/glog"
+	yaml "gopkg.in/yaml.v2"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/emptydirquota"
+)
+
+var (
+	volumeConfigKind       = "VolumeConfig"
+	volumeConfigAPIVersion = "kubelet.config.openshift.io/v1"
+)
+
+// Miror the TypeMeta from k8s since it doesn't have yaml tags.
+// We don't really use this right now anyway.  We just want
+// the on-disk format of the config to be right in case we
+// do want to version/parse the config in the k8s way later.
+type TypeMeta struct {
+	Kind       string `yaml:"kind"`
+	APIVersion string `yaml:"apiVersion"`
+}
+
+// VolumeConfig contains options for configuring volumes on the node.
+type VolumeConfig struct {
+	TypeMeta `yaml:"typeMeta,inline"`
+
+	// LocalQuota contains options for controlling local volume quota on the node.
+	LocalQuota LocalQuota `yaml:"localQuota"`
+}
+
+// LocalQuota contains options for controlling local volume quota on the node.
+type LocalQuota struct {
+	// perFSGroupInGiB can be specified to enable a quota on local storage use in GiB per unique FSGroup ID.
+	// At present this is only implemented for emptyDir volumes, and if the underlying
+	// volumeDirectory is on an XFS filesystem.
+	PerFSGroupInGiB int64 `yaml:"perFSGroupInGiB"`
+}
+
+// THIS IS PART OF AN OPENSHIFT CARRY PATCH
+// PatchVolumePluginsForLocalQuota checks if the node config specifies a local storage
+// perFSGroup quota, and if so will test that the volumeDirectory is on a
+// filesystem suitable for quota enforcement. If checks pass the k8s emptyDir
+// volume plugin will be replaced with a wrapper version which adds quota
+// functionality.
+func PatchVolumePluginsForLocalQuota(rootdir string, plugins *[]volume.VolumePlugin) error {
```

Comment:
can we get a test case for this?

or if not here, something that verifies that the function works moving forward as a follow-on?